### ISSUE TITLE
Some rules in .stylelintrc produces Undefined rule Error code 78

### ISSUE THREAD

**Issue by cjduncana** [olegskl/gulp-stylelint#18]
# TL;DR

Why does the rules `at-rule-name-case`, `function-name-case`, `property-case`, `selector-pseudo-element-case`, `selector-pseudo-class-case`, and `unit-case` produces an error?

# Problem

I have the following .styleintrc in the root of my project:

``` json
{
  "rules": {
    "at-rule-empty-line-before": [ "always", {
      "except": [ "blockless-group", "first-nested" ],
      "ignore": ["after-comment"],
    } ],
    "at-rule-name-case": "lower",
    "at-rule-semicolon-newline-after": "always",
    "block-closing-brace-newline-after": "always",
    "block-closing-brace-newline-before": "always-multi-line",
    "block-closing-brace-space-before": "always-single-line",
    "block-no-empty": true,
    "block-opening-brace-newline-after": "always-multi-line",
    "block-opening-brace-space-after": "always-single-line",
    "block-opening-brace-space-before": "always",
    "color-hex-case": "lower",
    "color-hex-length": "short",
    "color-no-invalid-hex": true,
    "comment-empty-line-before": [ "always", {
      "except": ["first-nested"],
      "ignore": ["stylelint-commands"],
    } ],
    "comment-whitespace-inside": "always",
    "declaration-bang-space-after": "never",
    "declaration-bang-space-before": "always",
    "declaration-block-no-ignored-properties": true,
    "declaration-block-no-shorthand-property-overrides": true,
    "declaration-block-semicolon-newline-after": "always-multi-line",
    "declaration-block-semicolon-space-after": "always-single-line",
    "declaration-block-semicolon-space-before": "never",
    "declaration-block-single-line-max-declarations": 1,
    "declaration-block-trailing-semicolon": "always",
    "declaration-colon-newline-after": "always-multi-line",
    "declaration-colon-space-after": "always-single-line",
    "declaration-colon-space-before": "never",
    "function-calc-no-unspaced-operator": true,
    "function-comma-newline-after": "always-multi-line",
    "function-comma-space-after": "always-single-line",
    "function-comma-space-before": "never",
    "function-linear-gradient-no-nonstandard-direction": true,
    "function-name-case": "lower",
    "function-parentheses-newline-inside": "always-multi-line",
    "function-parentheses-space-inside": "never-single-line",
    "function-whitespace-after": "always",
    "indentation": 4,
    "max-empty-lines": 1,
    "media-feature-colon-space-after": "always",
    "media-feature-colon-space-before": "never",
    "media-feature-no-missing-punctuation": true,
    "media-feature-range-operator-space-after": "always",
    "media-feature-range-operator-space-before": "always",
    "media-query-list-comma-newline-after": "always-multi-line",
    "media-query-list-comma-space-after": "always-single-line",
    "media-query-list-comma-space-before": "never",
    "media-query-parentheses-space-inside": "never",
    "no-eol-whitespace": true,
    "no-invalid-double-slash-comments": true,
    "no-missing-eof-newline": true,
    "number-leading-zero": "always",
    "number-no-trailing-zeros": true,
    "number-zero-length-no-unit": true,
    "property-case": "lower",
    "rule-non-nested-empty-line-before": [ "always-multi-line", {
      "ignore": ["after-comment"],
    } ],
    "selector-combinator-space-after": "always",
    "selector-combinator-space-before": "always",
    "selector-list-comma-newline-after": "always-multi-line",
    "selector-list-comma-space-before": "never",
    "selector-pseudo-element-case": "lower",
    "selector-pseudo-class-case": "lower",
    "selector-pseudo-element-colon-notation": "double",
    "selector-type-case": "lower",
    "string-no-newline": true,
    "unit-no-unknown": true,
    "unit-case": "lower",
    "value-list-comma-newline-after": "always-multi-line",
    "value-list-comma-space-after": "always-single-line",
    "value-list-comma-space-before": "never",
  },
}
```

When I run the Gulp task with stylelint, I get the following error:

```
Error in plugin 'gulp-stylelint'
Message:
    Undefined rule "at-rule-name-case"
Details:
    code: 78
```

I isolated that rule:

``` json
{
  "rules": {
    "at-rule-name-case": "lower",
  },
}
```

I get the same error:

```
Error in plugin 'gulp-stylelint'
Message:
    Undefined rule "at-rule-name-case"
Details:
    code: 78
```

I try to figure out what exactly is causing the failure. It does not seem to be the syntax, because I tested the plugin with other rules:

``` json
{
  "rules": {
    "at-rule-empty-line-before": [ "always", {
      "except": [ "blockless-group", "first-nested" ],
      "ignore": ["after-comment"],
    } ],
  },
}
```

``` json
{
  "rules": {
    "value-list-comma-newline-after": "always-multi-line",
    "value-list-comma-space-after": "always-single-line",
    "value-list-comma-space-before": "never",
  },
}
```

These rules run as expected. I return to my original rule set, and solely remove the rule in question. By doing that, another rule gave an error:

```
Error in plugin 'gulp-stylelint'
Message:
    Undefined rule "function-name-case"
Details:
    code: 78
```

I continued this process of removing the rules that gave me the error. The following are the error messages given by those rules inclusion:

```
Error in plugin 'gulp-stylelint'
Message:
    Undefined rule "property-case"
Details:
    code: 78
```

```
Error in plugin 'gulp-stylelint'
Message:
    Undefined rule "selector-pseudo-element-case"
Details:
    code: 78
```

```
Error in plugin 'gulp-stylelint'
Message:
    Undefined rule "selector-pseudo-class-case"
Details:
    code: 78
```

```
Error in plugin 'gulp-stylelint'
Message:
    Undefined rule "unit-case"
Details:
    code: 78
```

All of these rules have the value `lower`, so I thought that the error had to do with this. I checked the stylelint rules, and all of these failing rules can take `lower` as a value. Not only that, but there are other rules like `color-hex-case` and `selector-type-case` that have the value `lower` and do not produce an error.

Why does the rules `at-rule-name-case`, `function-name-case`, `property-case`, `selector-pseudo-element-case`, `selector-pseudo-class-case`, and `unit-case` produces the error `Undefined rule` when the other rules are capable of running fine?

**Comment by cjduncana** [comment]
The following is a functional version of the previous rule set with the problematic rules omitted:

``` json
{
  "rules": {
    "at-rule-empty-line-before": [ "always", {
      "except": [ "blockless-group", "first-nested" ],
      "ignore": ["after-comment"],
    } ],
    "at-rule-semicolon-newline-after": "always",
    "block-closing-brace-newline-after": "always",
    "block-closing-brace-newline-before": "always-multi-line",
    "block-closing-brace-space-before": "always-single-line",
    "block-no-empty": true,
    "block-opening-brace-newline-after": "always-multi-line",
    "block-opening-brace-space-after": "always-single-line",
    "block-opening-brace-space-before": "always",
    "color-hex-case": "lower",
    "color-hex-length": "short",
    "color-no-invalid-hex": true,
    "comment-empty-line-before": [ "always", {
      "except": ["first-nested"],
      "ignore": ["stylelint-commands"],
    } ],
    "comment-whitespace-inside": "always",
    "declaration-bang-space-after": "never",
    "declaration-bang-space-before": "always",
    "declaration-block-no-ignored-properties": true,
    "declaration-block-no-shorthand-property-overrides": true,
    "declaration-block-semicolon-newline-after": "always-multi-line",
    "declaration-block-semicolon-space-after": "always-single-line",
    "declaration-block-semicolon-space-before": "never",
    "declaration-block-single-line-max-declarations": 1,
    "declaration-block-trailing-semicolon": "always",
    "declaration-colon-newline-after": "always-multi-line",
    "declaration-colon-space-after": "always-single-line",
    "declaration-colon-space-before": "never",
    "function-calc-no-unspaced-operator": true,
    "function-comma-newline-after": "always-multi-line",
    "function-comma-space-after": "always-single-line",
    "function-comma-space-before": "never",
    "function-linear-gradient-no-nonstandard-direction": true,
    "function-parentheses-newline-inside": "always-multi-line",
    "function-parentheses-space-inside": "never-single-line",
    "function-whitespace-after": "always",
    "indentation": 4,
    "max-empty-lines": 1,
    "media-feature-colon-space-after": "always",
    "media-feature-colon-space-before": "never",
    "media-feature-no-missing-punctuation": true,
    "media-feature-range-operator-space-after": "always",
    "media-feature-range-operator-space-before": "always",
    "media-query-list-comma-newline-after": "always-multi-line",
    "media-query-list-comma-space-after": "always-single-line",
    "media-query-list-comma-space-before": "never",
    "media-query-parentheses-space-inside": "never",
    "no-eol-whitespace": true,
    "no-invalid-double-slash-comments": true,
    "no-missing-eof-newline": true,
    "number-leading-zero": "always",
    "number-no-trailing-zeros": true,
    "number-zero-length-no-unit": true,
    "rule-non-nested-empty-line-before": [ "always-multi-line", {
      "ignore": ["after-comment"],
    } ],
    "selector-combinator-space-after": "always",
    "selector-combinator-space-before": "always",
    "selector-list-comma-newline-after": "always-multi-line",
    "selector-list-comma-space-before": "never",
    "selector-pseudo-element-colon-notation": "double",
    "selector-type-case": "lower",
    "string-no-newline": true,
    "unit-no-unknown": true,
    "value-list-comma-newline-after": "always-multi-line",
    "value-list-comma-space-after": "always-single-line",
    "value-list-comma-space-before": "never",
  },
}
```

**Comment by cjduncana** [comment]
BTW, I am running gulp@3.9.1 and gulp-stylelint@1.1.0

**Comment by olegskl** [comment]
Hello and thanks for writing such a detailed issue.

The rules you mention have been released just a few days ago in [stylelint 6.0.0](https://github.com/stylelint/stylelint/blob/master/CHANGELOG.md#600). Errors are due to the fact that gulp-stylelint 1.x uses stylelint 5.x.

I will be releasing gulp-stylelint v2 that uses stylelint v6 pretty soon.

Related to #17

**Comment by cjduncana** [comment]
I did not realized this. Thanks for the prompt response. I'll be waiting for the new version.

**Comment by AntonioRedondo** [comment]
It looks like `gulp-stylelint` is again out of sync with the last release of `stylelint` 7.0.x. It's causing `undefined rules` errors with the below two rules:
- `"no-missing-end-of-source-newline"`
- `"media-feature-parentheses-space-inside"`

**Comment by olegskl** [comment]
Hello. Thanks for reporting it. I'll handle it tomorrow morning.

**Comment by olegskl** [comment]
Just released [v3.0.0 on npm](https://www.npmjs.com/package/gulp-stylelint) with styelint 7.x as a dependency.

**Comment by cyril-lamotte** [comment]
Hi, I have the same problème today :

`[09:24:47] Error in plugin 'gulp-stylelint'
Message:
    Undefined rule "number-zero-length-no-unit"
Details:
    code: 78
`

I guess I'm up to date :
`
$ npm gulp-stylelint -v
3.3.12
`

Do you now why ?
